### PR TITLE
Support for reading the Ambiata special recipe chord file

### PR DESCRIPTION
**Pull request by jystic**
This consumes the format produced by @amosr's work earlier today.

I'll join them together in a later PR.

### CODE REVIEW

File: data/sea/51-chord.h [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+    return (const idate_t *)entity_end;
+}
+
+static const ichord_t * INLINE ichord_next (const ichord_t *chord)
+{
+    const idate_t  *date_end = ichord_dates (chord) + chord->count;
+    const ichord_t *next     = (const ichord_t *)date_end;
+
+    return next->entity_size == 0 ? 0 : next;
```
Comment (amosr) [inline data/sea/51-chord.h]: I should probably explicitly filter out empty entity in the chord file output

File: data/sea/51-chord.h [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+static const idate_t * INLINE ichord_dates (const ichord_t *chord)
+{
+    size_t      entity0_size = (size_t) chord->entity_size + 1;
+    const char* entity_end   = chord->entity + entity0_size;
+
+    return (const idate_t *)entity_end;
+}
+
+static const ichord_t * INLINE ichord_next (const ichord_t *chord)
```
Comment (amosr) [inline data/sea/51-chord.h]: as we start writing more and more complicated C, I'd be tempted to start encoding invariants and preconditions as a Hoare logic sort of thing. doesn't have to be exact, but something like
```
{ chord /= 0 /\ VALID_CHORD_FILE }
chord' = ichord_next(chord)
{ chord' == 0 \/ chord'->entity > chord->entity }
```

just an idea for the future, I wouldn't worry about going back and adding them to this

File: src/Icicle/Sea/FromAvalanche/Psv.hs [hunk 3]
```diff
@@ -138,9 +133,9 @@ seaOfAllocProgram mode state
                 <> calloc "psv_max_row_count" (seaOfValType t)
 
    in vsep [ "/* " <> seaOfAttributeDesc (stateAttribute state) <> " */"
-           , programs <> " = " <> calloc maxChords stype
+           , programs <> " = " <> calloc "max_chord_count" stype
            , ""
-           , "for (iint_t ix = 0; ix < " <> maxChords <> "; ix++) {"
+           , "for (iint_t ix = 0; ix < max_chord_count; ix++) {"
```
Comment (amosr) [inline src/Icicle/Sea/FromAvalanche/Psv.hs]: oh wow, so the maximum chord count is totally dynamic? that's great

File: src/Icicle/Sea/FromAvalanche/Psv.hs [hunk 4]
```diff
@@ -148,21 +143,22 @@ seaOfAllocProgram mode state
...
  [ "#line 1 \"collect fleet state\""
  , "static void psv_collect_fleet (ifleet_t *fleet)"
  , "{"
- , "    imempool_t *into_pool   = imempool_create ();"
- , "    imempool_t *last_pool   = fleet->mempool;"
- , "    iint_t      chord_count = fleet->chord_count;"
+ , "    imempool_t *into_pool       = imempool_create ();"
+ , "    imempool_t *last_pool       = fleet->mempool;"
+ , "    iint_t      max_chord_count = fleet->max_chord_count;"
+ , "    iint_t      chord_count     = fleet->chord_count;"
```
Comment (amosr) [inline src/Icicle/Sea/FromAvalanche/Psv.hs]: ah. so this is the number of chords on the current entity - so we only need to garbage collect the first chord_count states?

**Comment by amosr** [comment]
superlative :sunflower:

File: data/sea/51-chord.h (hunk 2 above)
Comment (jystic) [inline data/sea/51-chord.h]: Yeah that's a great idea, I wonder if there's anything that can verify those kind of things?

File: src/Icicle/Sea/FromAvalanche/Psv.hs (hunk 3 above)
Comment (jystic) [inline src/Icicle/Sea/FromAvalanche/Psv.hs]: Yeah! It turned out to be easier to do it this way :)

File: src/Icicle/Sea/FromAvalanche/Psv.hs (hunk 4 above)
Comment (jystic) [inline src/Icicle/Sea/FromAvalanche/Psv.hs]: Yep that's right :)

File: data/sea/51-chord.h (hunk 2 above)
Comment (amosr) [inline data/sea/51-chord.h]: actually, there was a talk at SAPLING about "verifying C using Whiley". I just asked the guy, and he said "Hey, at the moment it's only really good for smallish array programs like the ones I did at the talk."
I think we'd have to hand-convert from Whiley to C too, which is a shame

**Pull request merged**